src/Workspaces/Core/Portable/Execution/AssetStorages.cs

```diff
@@ -156,13 +157,21 @@ public RemotableData GetRemotableData(PinnedRemotableDataScope scope, Checksum c
         public void RegisterSnapshot(PinnedRemotableDataScope snapshot, AssetStorages.Storage storage)
         {
             // duplicates are not allowed, there can be multiple snapshots to same solution, so no ref counting.
-            Contract.ThrowIfFalse(_storages.TryAdd(snapshot, storage));
+            if (!_storages.TryAdd(snapshot, storage))
+            {
+                // this should make failure more explicit
+                FailFast.OnFatalException(new Exception("who is adding same snapshot?"));
```

Comment:
Ultimately this ends up calling `Microsoft.CodeAnalysis.Environment.FailFast()` which only uses the exception for its `Message` property. So in practice it should be ok.